Framework/Muon/src/ApplyMuonDetectorsGrouping.cpp

```diff
@@ -0,0 +1,350 @@
+
+  Muon::DatasetParams params;
+  // don't fill in instrument, runs, periods.
+  params.label = m_groupedWS_name;
+  params.itemType = Muon::ItemType::Group;
+  params.itemName = m_groupName;
+  params.plotType = m_plotType;
+  params.version = 1;
+  const std::string wsName = generateWorkspaceName(params);
+  return wsName;
+}
+
+/*
+* Store the input properties as private member variables
+*/
+WorkspaceGroup_sptr
+ApplyMuonDetectorGrouping::getUserInput(Workspace_sptr &inputWS,
+                                        WorkspaceGroup_sptr &groupedWS) {
+
+  // Cast input WS to a grouping
+  auto muonWS = boost::make_shared<WorkspaceGroup>();
+  if (auto test = boost::dynamic_pointer_cast<MatrixWorkspace>(inputWS)) {
+    muonWS->addWorkspace(inputWS);
+  } else {
+    muonWS = boost::dynamic_pointer_cast<WorkspaceGroup>(inputWS);
+  }
+
+  m_groupedWS_name = groupedWS->getName();
+  m_wsName = inputWS->getName();
+  m_groupName = this->getProperty("groupName");
+  m_groups = this->getProperty("Grouping");
+  m_rebinArgs = this->getProperty("rebinArgs");
+  m_Xmin = this->getProperty("TimeMin");
+  m_Xmax = this->getProperty("TimeMax");
+  m_timeZero = this->getProperty("TimeZero");
+  m_loadedTimeZero = this->getProperty("TimeLoadZero");
+  m_summedPeriods = this->getProperty("SummedPeriods");
+  m_subtractedPeriods = this->getProperty("SubtractedPeriods");
+
+  std::string plotType;
+  plotType = this->getProperty("plotType");
+  if (plotType == "Counts") {
+    m_plotType = Muon::PlotType::Counts;
+  } else if (plotType == "Asymmetry") {
+    m_plotType = Muon::PlotType::Asymmetry;
+  } else {
+    // default to Counts.
+    m_plotType = Muon::PlotType::Counts;
+  }
+
+  return muonWS;
+}
+
+/**
+* Clip Xmin/Xmax to the range in the input WS
+*/
+void ApplyMuonDetectorGrouping::clipXRangeToWorkspace(
+    MatrixWorkspace_sptr &ws) {
+  double dataXMin;
+  double dataXMax;
+  ws->getXMinMax(dataXMin, dataXMax);
+  if (m_Xmin < dataXMin) {
+    m_Xmin = dataXMin;
+  }
+  if (m_Xmax > dataXMax) {
+    m_Xmax = dataXMax;
+  }
+}
+
+/**
+* Creates workspace, processing the data using the MuonProcess algorithm.
+*/
+Workspace_sptr
+ApplyMuonDetectorGrouping::createAnalysisWorkspace(Workspace_sptr &inputWS,
+                                                   bool isRaw) {
+
+  // Store all the options for Muon Process
+  Grouping grouping;
+  grouping.description = "no description";
+  grouping.groupNames.emplace_back(m_groupName);
+  grouping.groups.emplace_back(m_groups);
+
+  Muon::AnalysisOptions options(grouping);
+
+  options.summedPeriods = m_summedPeriods;
+  options.subtractedPeriods = m_subtractedPeriods;
+  options.timeZero = m_timeZero;
+  options.loadedTimeZero = m_loadedTimeZero;
+  options.timeLimits.first = m_Xmin;
+  options.timeLimits.second = m_Xmax;
+  options.rebinArgs = isRaw ? "" : m_rebinArgs;
+  options.plotType = m_plotType;
+  options.groupPairName = m_groupName;
+
+  IAlgorithm_sptr alg =
```

Comment:
Is it worth considering having a method `runMuonProcess` that ran the algorithm and returned the workspace that could encapsulate the code from here to the end of this function? It would only have to accept a `MuonAnalysisOptions` type.